### DISCUSSION TITLE
CSS Frameworks

### DISCUSSION THREAD

**Discussion by kcramakrishna** [Digital-Assistant/Digital_Assistant_Client#341] [Ideas]
We should upgrade to SCSS and update to a better css framework too - either latest bootstrap (custom namespace to avoid conflict with customer/target web app.)

Move from CSS to SCSS (SASS):
--------------------------------
https://www.geeksforgeeks.org/what-is-the-difference-between-css-and-scss/

Is bootstrap the right solution or should we consider alternatives like tailwind:
-----------------------------------------------------------------------------
https://www.spec-india.com/blog/bootstrap-alternatives
https://tailwindcss.com/

**Comment by kcramakrishna** [comment]
Some articles on Tailwind vs. bootstrap:
https://www.hallme.com/blog/choosing-the-best-css-framework-for-your-site-bootstrap-vs-tailwind-css/
https://www.geeksforgeeks.org/tailwind-css-vs-bootstrap/
https://blog.logrocket.com/tailwind-css-vs-bootstrap-ui-kits/

Looks like the consensus is that:
1. Tailwind helps you design very custom look/feel while bootstrap is more standardised. 
2. Tailwind is smaller in size BUT is younger while bootstrap has much better community. 

Leaning towards Bootstrap for now but will research more.

**Comment by kcramakrishna** [comment]
https://www.sitepoint.com/bootstrap-5-new-features-examples/
https://www.npmjs.com/package/bootstrap-sass-namespace
https://jonasjancarik.medium.com/namespacing-bootstrap-css-419961f4d1f8
https://www.youtube.com/watch?v=f4fikVsxbzw
https://wiki.izenda.us/FAQ/bootstrap-namespacing
https://blog.wplauncher.com/how-to-namespace-bootstrap/

https://formden.com/blog/isolate-bootstrap

**Comment by kcramakrishna** [comment]
https://docs.oracle.com/cd/F22607_01/reference-api/Bootstrap.html
https://developer.mozilla.org/en-US/docs/Web/CSS/@namespace
https://www.w3.org/community/webed/wiki/CSS/Namespaces
https://www.geeksforgeeks.org/javascript-namespace/

**Comment by lakshmanveti** [comment]
I would say bootstrap is good enough for regular web applications(Non componentized), but for applications like react, angular and vue need more adoptable components, one of the famous open source UI Library for react is [ANTD](https://ant.design/
) . It has plethora of easy-to-use components that are useful for building elegant user interfaces which minimizes development effort and testing, besides the components, it also supports scaffolding , theming etc.

**Reply by kcramakrishna** [reply]
Looks very interesting. Is this really free? I.e. will we be forced into a closed source version at any time? 
 Any other risks?

What are the alternatives and competing projects?

**Reply by lakshmanveti** [reply]
Yes it is free, and we already used it for one of the biggest low-code/no-code platform vendor.
More info [here](https://github.com/ant-design/ant-design/blob/master/LICENSE)

**Reply by nagamunemma** [reply]
Apart from Ant D Material UI also there. But Ant D is better than Material UI. Here are below links to know more frameworks.

https://www.specbee.com/blogs/how-to-customize-react-theme-using-ant-design#:~:text=Created%20by%20Alibaba%2C%20the%20Ant,overtaken%20them%20with%2077.5k!

https://stackshare.io/ant-design

https://blog.logrocket.com/comparing-popular-react-component-libraries/

**Reply by kcramakrishna** [reply]
Agreed. Please use AntD.

**Comment by kcramakrishna** [comment]
This was suggested by someone and looks like a promising approach to implement.
Shadow DOM:
https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM 
https://developer.mozilla.org/en-US/docs/Web/Web_Components
https://web.dev/shadowdom-v1/
https://javascript.info/shadow-dom
https://stackoverflow.com/questions/41546873/is-it-possible-to-use-shadow-dom-in-a-webworker

Browser compatibility of ShadowDOM:
https://www.lambdatest.com/web-technologies/shadowdomv1 (as on 30-June-2022)

Note on performance:
https://nolanlawson.com/2022/06/22/style-scoping-versus-shadow-dom-which-is-fastest/ 

From W3.org
https://www.w3.org/TR/shadow-dom/ 

Need to dig deeper and see if implementing this makes sense. 

https://caniuse.com/shadowdomv1

**Comment by kcramakrishna** [comment]
Upgrading from AntD 4.1.x to AntD 5.x 
https://ant.design/docs/react/migration-v5

**Comment by kcramakrishna** [comment]
While we have decided to use AntD and the current version uses AntD, we can revisit "AntD vs. bootstrap5"

Argument is that way more people know bootstrap compared to AntD and hence can be easier to find people to maintain.

Comparison:
https://npmcompare.com/compare/antd,bootstrap,material-ui,semantic-ui-react
Size:
antd is 379KB and bootstrap is 15KB - https://npmtrends.com/ant-design-vs-antd-vs-bootstrap-vs-reactstrap 

Trends: bootstrap is more than twice as popular. 
https://npmtrends.com/ant-design-vs-antd-vs-bootstrap-vs-core-ui-vs-material-ui-vs-react-bootstrap-vs-reactstrap-vs-semantic-ui
https://npmtrends.com/ant-design-vs-antd-vs-bootstrap-vs-core-ui-vs-material-ui-vs-react-bootstrap-vs-reactstrap-vs-semantic-ui

This is not urgent. We can debate this a bit more.

**Comment by kcramakrishna** [comment]
Did some in-depth reading about antd and its problems. 

1. The biggest problem seems to be that a good chunk of the documentation is in Chinese:
https://old.reddit.com/r/reactjs/comments/sxwyad/any_reason_why_not_to_use_ant_design_in_2022/
See comments here: https://www.youtube.com/watch?v=IEqmSROj5Uc 

2. It seems to be a bit bloated:
https://orkhanhuseyn.medium.com/ant-design-or-elephant-design-a03c19549553 

3. Has some performance issues:
https://youtu.be/IEqmSROj5Uc?t=235

**Reply by kcramakrishna** [reply]
Maybe consider lates bootstrap version 5.x 
https://www.geeksforgeeks.org/difference-between-bootstrap-4-and-bootstrap-5/
